src/lib.rs

```diff
@@ -287,69 +327,518 @@ impl<D: AsMut<[Digit]>> BigInt<D> {
+    }
+}
+
+impl<D: AsMut<[Digit]>> BigInt<D> {
+    /// Computes `a * b` where `b` is a digit, writes the result to `self`.
+    ///
+    /// It is a low-level function allowing you to manage allocations. Its safe analogue is
+    /// [`&a * b`](#impl-Mul%3CDigit%3E)
+    ///
+    /// ## Safety
+    /// Assuming that `a` has [size] `n`:
+    /// * `n > 0`
+    /// * [Size][size] of `self` must be exactly `n+1` limbs
+    /// * Size of `scratch_space` must be exactly
+    /// [`ScratchSpace::multiplication(n, 1)`](ScratchSpace::multiplication) limbs
+    ///
+    /// [size]: BigInt::size
+    ///
+    /// ## Constant time
+    /// This function is constant time in size of its arguments, ie. number of instructions and
+    /// memory access patterns remain the same for the same `n`.
+    pub unsafe fn mul_digit_unchecked(
+        &mut self,
+        a: &BigInt<impl AsRef<[Digit]>>,
+        b: Digit,
+        mut scratch_space: ScratchSpace,
+    ) {
+        self.mul_unchecked(a, &BigInt::from_digits(&[b]), scratch_space);
     }
 }
 
-impl<'n, D: AsRef<[Digit]>> Add<&'n BigInt<D>> for Digit {
+impl_digit_operation!(Mul, mul, u8);
+impl_digit_operation_reverse!(Mul, mul, u8);
+impl_digit_operation!(Mul, mul, u16);
+impl_digit_operation_reverse!(Mul, mul, u16);
+impl_digit_operation!(Mul, mul, u32);
+impl_digit_operation_reverse!(Mul, mul, u32);
+impl_u64_operation!(Mul, mul);
+impl_digit_operation_reverse!(Mul, mul, u64);
+impl_digit_operation_reverse!(Mul, mul, Digit);
+
+impl<D1, D2> Div<&BigInt<D2>> for &BigInt<D1>
+where
+    D1: AsRef<[Digit]>,
+    D2: AsRef<[Digit]>,
+{
     type Output = BigInt;
 
-    #[inline(always)]
-    fn add(self, rhs: &BigInt<D>) -> Self::Output {
-        rhs + self
+    /// Computes `floor(self / rhs)`
+    ///
+    /// ## Constant time
+    /// Division is constant time in size of its arguments. It means that the number of
+    /// instructions and memory access patterns remain the same for the same
+    /// `self.size()` and `rhs.size()` (provided that `rhs` does not start from 0).
+    ///
+    /// ## Zeroizing
+    /// This function might allocate some auxiliary buffer on the heap.
+    /// Before deallocating, it gets zeroed out.
+    ///
+    /// ## Complexity
+    /// The complexity is always `O(self.size() * rhs.size())`.
+    ///
+    /// ## Panics
+    /// The method panics if `rhs` equals zero.
+    fn div(self, rhs: &BigInt<D2>) -> Self::Output {
+        let n = self.size();
+        // we have to clone `self` in order to be able to use unchecked division,
+        // as it requires its first argument to be mutable
+        let mut a = self.reallocate();
+        let (a, b) = (a.as_mut(), rhs.as_ref());
+
+        // m is the length of b without leading zeroes
+        let m = ops::len_without_leading_zeroes(b);
+
+        if m == 0 {
+            panic!("Division by zero");
+        }
+
+        // if m > n, then definitely b > a and so floor(a/b) = 0.
+        if m > n {
```

Comment:
We are allowed to use branches (`if ... { ... }` constructions) that depend only on size of input arguments (since we claim that we constant-time in size of arguments). Strictly speaking, this branch depend on *value* of `rhs`, since `m` is length of `b` without leading zeroes. So (again) strictly speaking, we cannot afford having this branch. 

Do you see any ways to remove this check?

Reply:
Ah I see you already pointed out this place. I'll try to figure something out.

Reply:
In general, I'm not sure how much effort should be put into guaranteeing constant time for the case of leading zeroes. Do you think it's important and we should sacrifice performance by e.g. making `remove_leading_zeroes` constant time?

Reply:
Imo we should. It's a point of crate to be less efficient but more secure. But in above comment I'm saying that even if we make `remove_leading_zeroes` constant time, the whole division won't become constant-time.

Reply:
Right. I don't see a way to enforce constant-time here if divisor is allowed to be zero-padded tbh.